### PR TITLE
Optimize the transformRect and transformPoint methods in matrix_utils.

### PR DESCRIPTION
**Pull request by flar**
Primarily these methods no longer allocate any objects other than their
return values.

Additionally, the math in the methods is reduced compared to the general
case math based on known input conditions.

Results from a Moto E2 phone...
Before:
```
I/flutter (12908): MatrixUtils.transformRectPerspective: 3167.3 ns per iteration
I/flutter (12908): MatrixUtils.transformRectAffine: 3098.9 ns per iteration
I/flutter (12908): MatrixUtils.TransformPointPerspective: 640.2 ns per iteration
I/flutter (12908): MatrixUtils.TransformPointAffine: 633.0 ns per iteration
```
After:
```
I/flutter ( 8339): MatrixUtils.transformRectPerspective: 1026.5 ns per iteration
I/flutter ( 8339): MatrixUtils.transformRectAffine: 647.7 ns per iteration
I/flutter ( 8339): MatrixUtils.transformPointPerspective: 511.9 ns per iteration
I/flutter ( 8339): MatrixUtils.transformPointAffine: 477.4 ns per iteration
```

Results from a Samsung Note 9...
Before
```
I/flutter ( 3761): MatrixUtils.transformRectPerspective: 432.0 ns per iteration
I/flutter ( 3761): MatrixUtils.transformRectAffine: 430.4 ns per iteration
I/flutter ( 3761): MatrixUtils.transformPointPerspective: 64.6 ns per iteration
I/flutter ( 3761): MatrixUtils.transformPointAffine: 63.5 ns per iteration
```
After:
```
I/flutter ( 5862): MatrixUtils.transformRectPerspective: 80.4 ns per iteration
I/flutter ( 5862): MatrixUtils.transformRectAffine: 55.0 ns per iteration
I/flutter ( 5862): MatrixUtils.transformPointPerspective: 39.4 ns per iteration
I/flutter ( 5862): MatrixUtils.transformPointAffine: 39.4 ns per iteration
```

### CODE REVIEW

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (dnfield) [inline packages/flutter/test/painting/matrix_utils_test.dart]: nit: these names are a little hard to read and we try to avoid abbreviations.  Maybe this one can just be `tenByTwenty`, and the other one `fourtyByFourty`?

Alternatively, you could name them based on the expected inputs and outputs, even as simple as `inputRect` and `outputRect`.

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (dnfield) [inline packages/flutter/test/painting/matrix_utils_test.dart]: Perhaps it would give us a nice sanity check to validate the results of this against the implementations from vector_math.  I trust that you did the right thing here, but it would make reviewing/changing this easier.

We might also want to consider upstreaming this into vector_math.

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (dnfield) [inline packages/flutter/lib/src/painting/matrix_utils.dart]: These and similar have to be split up on multiple lines I believe.

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: I called it rectangle20x20 and inlined the other rect in the one place it was used (as an expected answer for one of the tests). Or are words preferred to numbers?

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: _vectorWiseTransformRect used in the other (perspective) test uses vector_math. So, which of the following would make sense:

1 - MatrixUtils vs. _vectorWise/aka vector_math
2 - MatrixUtils vs. manually computed answers
3 - _vectorWise vs. manually computed answers

We currently do #2, are you suggesting replacing it with #1, or adding #1 (or #3)?

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: Dang it. I saw the warnings in Android Studio, but was going to come back and see if I could find some other format that would make the logic as easy to follow but comply. Unfortunately, this formatting does a great job of pointing out that we are deciding which "wall" to add the vector component to and splitting it up just wastes lines of code to obscure that logic.

`(wx < 0) ? left += wx : right += wx;`
works fine, passes the analyzer, is uglier than what I wrote and slightly misleading in that the value is discarded, but at least maintains making the logic visible.

Grrr...

Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: (And it would be interesting if we could write `(wx < 0 ? left : right) += wx;` but ternary operators can't be LHS of an assignment.)

Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: Another alternative:
```
      double left  = rx + (wx > 0 ? 0 : wx) + (hx > 0 ? 0 : hx);
      double right = rx + (wx < 0 ? 0 : wx) + (hx < 0 ? 0 : hx);
```
But it leads to more work unless the compiler is smart about how it evaluates the ternary expressions and I don't think the logic is as clear.

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (dnfield) [reply packages/flutter/test/painting/matrix_utils_test.dart]: Numbers are fine, we just try to avoid abbreviations.

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (dnfield) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: If it were me I'd just format these on separate lines rather than trying to come up with a ternary that works with this, but as it is right now it won't pass.

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (dnfield) [inline packages/flutter/test/painting/matrix_utils_test.dart]: Ah! I missed this.

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (dnfield) [reply packages/flutter/test/painting/matrix_utils_test.dart]: Sorry, I missed where you were doing this.  Resolved.

Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: This comment is in a test file but you mention "upstreaming this into vector_math". Do you mean the test or the optimization (or both)?

The problem with moving the optimization upstream is that we use different classes to store our data than vector_math (except for Matrix4) so to avoid unpacking/repacking/moving the data to a different object, ...? This is why I inlined these calculations to avoid transfer of data for a common operation.

Note that vector_math has its own bounding objects, but they are much more abstract and complicated than our Rect.

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: Does this resolve the comment you made about comparing the test output to vector_math?

Comment (dnfield) [reply packages/flutter/test/painting/matrix_utils_test.dart]: Yes.

**Comment by flar** [comment]
Newest update resolves:
- Expanded left/right/top/bottom += whxy logic to multi-line to fix analyzer issues
- Fixed variable names in test

I did not add any code to compare test results to vector_math

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: That's what I did, but I grumbled a bit as I dit it. ;)

File: packages/flutter/test/painting/matrix_utils_test.dart
Comment (dnfield) [reply packages/flutter/test/painting/matrix_utils_test.dart]: I don't see upstreaming as a blocker to this PR - more that this approach seems to be an improvement, and it might be something that would be valuable to have available via vector_math rather than burried in Flutter.

I think we can land this without doing that - if I'm not missing something here, it would be enough to just open a bug on vector_math to track making this available there.

Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: I did think of this as I was doing it and I did look through the code in vector_math, but in the end I realized that our specific needs aren't a great match for the tone of their API. In particular, the bounds objects they provide feel more oriented to the kind of math that a 3D CAD or game package would require and doesn't seem to speak the language of rectangles laid out on a screen.

Another thing to consider is how big vector_math is and how good our tree shaking is. Vector4, for example, has what looks like several hundred properties on it. Do we end up with getter code for all of those properties in a flutter app that is just trying to figure out which cat video someone poked with their finger?

This is a larger discussion than just this one fix. I'd almost lean towards creating our own matrix package that is more streamlined and tailored to our needs, but I believe we have Matrix4 baked into our APIs?

Comment (dnfield) [reply packages/flutter/test/painting/matrix_utils_test.dart]: From what I understand, vector_math _is_ the package we created to have matrix math available in Flutter, though it's certainly got other uses as well and has some other large contributions outside of what Flutter needs.

Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: Hmmm, well it doesn't provide any Rect-like object so this would require more than just adding a new method. And then we deal with data ferrying between our Rect and its Rect or we make a breaking change and move our Rect to the vector_math package?

Comment (dnfield) [reply packages/flutter/test/painting/matrix_utils_test.dart]: We can't move our Rect, that's defined in dart:ui which isn't allowed to have dependencies.

I'm not sure if there's a really good way to do this, but as I said It shouldn't block this PR from landing either way.

Comment (flar) [reply packages/flutter/test/painting/matrix_utils_test.dart]: I started at looking at how hard it would be to do just the Point version of this inside vector_math and came away not happy with the state of its APIs (consistency, capability, etc.).

Sad bear... :(

**Review by dnfield — APPROVED** [review]
LGTM - I would defer to yjbanov though

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (yjbanov) [inline packages/flutter/lib/src/painting/matrix_utils.dart]: I'd expect `math.min` to be inlined and intrisified, but this should work too.

**Review by yjbanov — APPROVED** [review]

**Review by liyuqian — APPROVED** [review]
LGTM!

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: OK, that explains some of the backwards code I saw in the vector_math package.

(I'm used to Java which would have handled this differently.)

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: I got a minor improvement out of this. It wouldn't cost a lot to remove it if it looks confusing. It would be nice if the Dart math package provided min3 and min4 for us.

A tree of comparisons was actually the most optimal, but was way too wordy for being just a helper function for this more complicated operation in a package that wasn't designed to be a general math facility. This was the best blend of being simple code and almost as fast as the completely unrolled version.

I'll also point out that I optimized this at a time when the non-perspective case was using it, but then I later realized that there was an even simpler way to construct the bounds of a delta parallelogram and now only the rarer and longer perspective case uses this code.

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: I'm not sure this makes sense, but I implemented both of these "range check" proposals and they slowed it down on both of my test devices.  I got the following results:

Moto E2:
```
I/flutter ( 5449): MatrixUtils.transformRectPerspective: 1038.7 ns per iteration
I/flutter ( 5449): MatrixUtils.transformRectAffine: 663.2 ns per iteration
I/flutter ( 5449): MatrixUtils.transformPointPerspective: 514.4 ns per iteration
I/flutter ( 5449): MatrixUtils.transformPointAffine: 489.8 ns per iteration
```

Samsung Note 9:
```
I/flutter (14052): MatrixUtils.transformRectPerspective: 86.3 ns per iteration
I/flutter (14052): MatrixUtils.transformRectAffine: 56.2 ns per iteration
I/flutter (14052): MatrixUtils.transformPointPerspective: 44.1 ns per iteration
I/flutter (14052): MatrixUtils.transformPointAffine: 41.6 ns per iteration
```

I ran it multiple times to see if they were flukes, but they never were as fast as the runs I did on the last commit. :(

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: See above. Resolving this as a dup.

File: packages/flutter/lib/src/painting/matrix_utils.dart
Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: To be specific, here is the actual change I made to put a reference to storage[15] first in both methods:

```
diff --git a/packages/flutter/lib/src/painting/matrix_utils.dart b/packages/flutter/lib/src/painting/matrix_utils.dart
index c6da1c550..55b5ab742 100644
--- a/packages/flutter/lib/src/painting/matrix_utils.dart
+++ b/packages/flutter/lib/src/painting/matrix_utils.dart
@@ -131,9 +131,9 @@ class MatrixUtils {
     // dropping the resulting Z coordinate, and normalizing only
     // if needed.
 
-    final double rx = storage[0] * x + storage[4] * y + storage[12];
-    final double ry = storage[1] * x + storage[5] * y + storage[13];
-    final double rw = storage[3] * x + storage[7] * y + storage[15];
+    final double rw = storage[15] + storage[3] * x + storage[7] * y;
+    final double rx = storage[12] + storage[0] * x + storage[4] * y;
+    final double ry = storage[13] + storage[1] * x + storage[5] * y;
     if (rw == 1.0) {
       return Offset(rx, ry);
     } else {
@@ -301,6 +301,10 @@ class MatrixUtils {
     // optimized perspective = 12 mul + 39 add/sub/compare
     // optimized affine      =  8 mul + 17 add/sub/compare
 
+    final bool isAffine = storage[15] == 1.0 &&
+                          storage[ 3] == 0.0 &&
+                          storage[ 7] == 0.0;
+
     final double wx = storage[0] * w;
     final double hx = storage[4] * h;
     final double rx = storage[0] * x + storage[4] * y + storage[12];
@@ -309,7 +313,7 @@ class MatrixUtils {
     final double hy = storage[5] * h;
     final double ry = storage[1] * x + storage[5] * y + storage[13];
 
-    if (storage[3] == 0.0 && storage[7] == 0.0 && storage[15] == 1.0) {
+    if (isAffine) {
       double left  = rx;
       double right = rx;
       if (wx < 0) {
```

Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: I tried this again with back to back runs on both devices with and without that patch and the benchmark modified to use a lot more runs and the timings I got with the proposed commit were faster than the timings I got with the additional patch in the previous comment in most cases. The cases where changing the order showed a win were isolated and slim. The cases where leaving it as submitted showed a win were more significant and more numerous and the averages were in favor of the fix as submitted.  This is really in the noise in the end so I'm going to leave it as I've already submitted it.

https://docs.google.com/spreadsheets/d/1FhOgYeEfpY52iUQiUGvx3xWi4ROd2HuGyIvsy1lgqms/edit?usp=sharing

Comment (yjbanov) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: /cc @mraleph, who could shed some light on what's going on behind the scenes. I didn't expect a huge improvement from eliminating range checks, but I certainly didn't expect a slow down 😕 

**tl;dr** an optimization we tried is to read the last index out of a `Float64List` first so that the compiler only range checks once and lets smaller indices bypass the range check, i.e.:

```
final a = ...;
// The following should be faster
a[15]; a[10]; a[5]; a[0];
// than this
a[0]; a[5]; a[10]; a[15];
```

Because in the first case if `a[15]` doesn't throw a range error, the following statements cannot possibly throw due to smaller indices, so a range check isn't necessary. But @flar is observing a _slow down_.

Comment (mraleph) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: The slow down most likely originated from creating boolean and the dispatching on the boolean `if (isAffine) {` part of the change. Materialising boolean and dispatching on it later is more expensive than the chain of comparisons. 

... but I also discovered that we have a regression in our bounds check elimination - we no longer would eliminate bounds checks in a sequence like what @yjbanov written. We used to be able to do that, but that got broken - and we did not have tests for it. I filed https://github.com/dart-lang/sdk/issues/37687 to address that. 

[Benefits of eliminating bounds checks are pretty minimal though especially on the code where the performance is dominated by memory traffic and floating point math]

Comment (flar) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: A question on bounds check elimination...

If the code had `if (a.length < 16) return null;` towards the top (with the understanding that it would never happen so returning null is mostly vestigial), would the compiler be able to eliminate all bounds checking for constant integer indices < 16?

Comment (mraleph) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: The answer to that question is complicated... Theoretically `if (a.length < 16) return null;` should have the same effect as just loading `a[15]`... And in JIT it would, but in practice it is broken right now too :-/

Comment (mraleph) [reply packages/flutter/lib/src/painting/matrix_utils.dart]: (`if (a.length < 16) throw 'impossible';` would be better because it does not polute type information about the return value).

**Pull request merged**
